### ISSUE TITLE
Feature Request: Weight Normalization

### ISSUE THREAD

**Issue by Skylion007** [ppwwyyxx/tensorpack#344]
It would be nice to have Weight Normalization implemented from [this paper](https://arxiv.org/pdf/1602.07868.pdf) if it's not already. The technique has also been shown to have various other applications in [GANs](https://arxiv.org/pdf/1704.03971.pdf). It seems like it would complement the l2 and l1 regularizers well.

**Comment by ppwwyyxx** [comment]
We do not focus on models/symbolic functions and don't intend to add any more layers. We will shift the existing ones to `tf.layers` instead. You should be able to use any symbolic functions in tensorflow for your model.

In particular, `varreplace.remap_variables` will be a helpful function to write weight normalization.

**Comment by Skylion007** [comment]
I am really curious how the regularizer will change since it looks like most tf.layers have a regularizer  as an argument? That seems like it would be more difficult to replace with simple function decorators. I definitely encourage using tf.layers for consistency, but I just want to make sure any regularizer I write will be future proof.

**Comment by ppwwyyxx** [comment]
You can use either regularizers of tf.layers, or the existing `regularize_cost()` function -- it's just a symbolic function and you can call it any time anywhere when it suits.
